In [1]:
import matplotlib.pyplot as plt
import sys

Consider the following test graphs:
- http://snap.stanford.edu/data/email-Eu-core.html
- http://snap.stanford.edu/data/com-Amazon.html
- http://snap.stanford.edu/data/com-LiveJournal.html
- http://snap.stanford.edu/data/com-Orkut.html
- (https://snap.stanford.edu/data/com-Friendster.html)

It can also be usefull to consider some toy graphs (e.g. manually created graphs with a dozen nodes) to test your programs.

## Exercise 1: BFS

Implement an efficient BFS algorithm and test it on the four graphs.  
Use your BFS implementation to compute a good lower (and upper) bound to the diameter of a graph.
Test your program on the 4 downloaded graphs and for each one of them report the lower bound you have found as well as the running time of your program.

# Chargement du graphe en memoire

In [4]:
def load_graph(file):
    f = open(file, 'r')
    contenu=f.read() 
    Liste=contenu.split( )
    d={}
    for l in Liste:
        d[l]=[]
    j=0   
    for j in range(len(Liste)-1):
             for k in d.keys():
                   if k==Liste[j]:
                        if j%2==0:
                            d[k].append(Liste[j+1])
                        if j%2!=0:
                            d[k].append(Liste[j-1])
    f.close()
    return(d)


# les étapes à suivre pour calculer la borne suppérieure et inférieure d'un graphe
- On choisit aléatoirement un sommet s.
- On détermine à l’aide d’un BFS un sommet x le plus éloigné possible de s.
- On détermine à l’aide d’un BFS un sommet y le plus éloigné possible de x.
- On renvoie la distance d(x, y).
- d(x, y) <= diam(graph) <= 2d(x, y).
- On importe la fonction time qui nous permet de calculer le temps d'éxecution du BFS



In [ ]:
from datetime import datetime
start_time =datetime.now()
def bfs_algo(graph,s): 
    
    queue=[s]
    dist={}
    visited=[]

    dist[s]=0
    
    while queue:
        node=queue.pop(0)
        visited.append(node)
        for neighbour in graph[node]:
            if neighbour not in visited:
                visited.append(neighbour)
                dist[neighbour]=dist[node]+1
                queue.append(neighbour)
    
    print("le noeud X le plus loin de la source S est le: ",visited[len(visited)-1],"avec une distance de :", dist[visited[len(visited)-1]] )
print(bfs_algo(load_graph("inet.txt"),"0"))
end_time=datetime.now()
print('le temps d éxecution du BFS1 est: {}'.format(end_time - start_time))


def bfs_algo2(graph,s):
    queue=[s]
    dist={}
    visited=[]

    dist[s]=0
    
    while queue:
        node=queue.pop(0)
        visited.append(node)
        for neighbour in graph[node]:
            if neighbour not in visited:
                visited.append(neighbour)
                dist[neighbour]=dist[node]+1
                queue.append(neighbour)
   
    print("le diamètre de ce graphe est :", dist[visited[len(visited)-1]] )

print("veuillez saisir le noeud plus loin de la source trouvé à l'aide du BFS1")
X=input()
print(bfs_algo2(load_graph("inet.txt"),X))
   


le noeud X le plus loin de la source S est le:  1713 avec une distance de : 24
None
le temps d éxecution du BFS1 est: 0:01:03.279662
veuillez saisir le noeud plus loin de la source trouvé à l'aide du BFS1


# Récapitulatif 
- email-eu-core.txt: temps d'exécution= 3s / lower bound=6 / upper bound=12
- inet.txt: temps: 1 min:03 sec / lower bound=24 /upper bound=48
- amazon: temps d'éxecution très long( plus de 15 min avec un core i5 et une connexion à distance)

# Exercise 2: Triangles

Implement an efficient algorithm for counting triangles. Test your program on the 4 downloaded graphs and report the number of triangles as well as the running time of your program.

Generalize your program to compute the transitivity ratio and the clustering of the graph. Test your program on the 4 downloaded graphs and report the two values as well as the running time of your program.

In [22]:
from datetime import datetime
start_time =datetime.now()

def adjList_to_adjMatrix(graph):
    keys=sorted(graph.keys())
    size=len(keys)
    M = [ [0]*size for i in range(size) ]
    for a,b in [(keys.index(a), keys.index(b)) for a, row in graph.items() for b in row]:
           M[a][b] = 2 if (a==b) else 1
    return M

def countTriangle(file):
    adjList=load_graph(file)
    adjMatrix=adjList_to_adjMatrix(adjList)
    nodes = len(adjMatrix) 
    count_Triangle = 0  
    for i in range(nodes): 
        for j in range(nodes): 
            for k in range(nodes): 
                if( i!=j and i !=k and j !=k and 
                        adjMatrix[i][j] and adjMatrix[j][k] and adjMatrix[k][i]): 
                    count_Triangle += 1
    return int (count_Triangle/6)
print("le nombre de triangles est :", countTriangle("email-Eu-core.txt"))
end_time=datetime.now()
print('le temps d éxecution de ce programme est : {}'.format(end_time - start_time))

le nombre de triangles est : 105461
le temps d éxecution de ce programme est : 0:03:28.531289


#### Résultats:
- email-core.txt: temps d'éxection=3 min et 28 sec / nombre de triangles: 105461
- inet.txt: temps d'attente trop important

## Exercice 3 (bonus): Bron-Kerbosch

Implement an efficient Bron-Kerbosch algorithm following the wikipedia pseudocode: https://en.wikipedia.org/wiki/Bron%E2%80%93Kerbosch_algorithm. Test your program on the 4 graphs.
Report the number of maximal cliques for each graphs as well as the running time of your program.

ssdqs


Plot the number of maximal cliques of size k (for k from 1 to the size of a maximum clique) as a function of k for each graph.